src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -49,42 +49,126 @@ public class WorkitemOrchestrator : IWorkitemOrchestrator
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(workitemMetadata, nameof(workitemMetadata));
+
+            if (workitemMetadata.Status != WorkitemStoreStatus.ReadWrite)
+            {
+                return;
+            }
+
+            (long CurrentWatermark, long NextWatermark)? watermarkEntry = null;
+            WorkitemInstanceIdentifier identifier = null;
+
+            try
+            {
+                watermarkEntry = await _indexWorkitemStore
+                    .GetCurrentAndNextWorkitemWatermarkAsync(workitemMetadata.WorkitemKey, cancellationToken)
+                    .ConfigureAwait(false);
+
+                // Get the workitem from blob store
+                identifier = new WorkitemInstanceIdentifier(
+                    workitemMetadata.WorkitemUid,
+                    workitemMetadata.WorkitemKey,
+                    workitemMetadata.PartitionKey,
+                    watermarkEntry.Value.CurrentWatermark);
+                var storeDicomDataset = await GetWorkitemBlobAsync(identifier, cancellationToken).ConfigureAwait(false);
+
+                // update the procedure step state
+                var updatedDicomDataset = new DicomDataset(storeDicomDataset);
+
+                // Add/Update the procedure step state in the blob
+                var targetProcedureStepStateStringValue = targetProcedureStepState.GetStringValue();
+                updatedDicomDataset.AddOrUpdate(DicomTag.ProcedureStepState, targetProcedureStepStateStringValue);
+
+                // if there is a reason for cancellation, set it in the blob
+                dataset.TryGetString(DicomTag.ReasonForCancellation, out var cancellationReason);
+                if (!string.IsNullOrWhiteSpace(cancellationReason))
+                {
+                    updatedDicomDataset.AddOrUpdate(DicomTag.ReasonForCancellation, cancellationReason);
+                }
+
+                // store the blob with the new watermark
+                await StoreWorkitemBlobAsync(identifier, updatedDicomDataset, watermarkEntry.Value.NextWatermark, cancellationToken)
                     .ConfigureAwait(false);
 
-                await _indexWorkitemStore.EndAddWorkitemAsync(partitionKey, identifier.WorkitemKey, cancellationToken);
+                // Update the workitem procedure step state in the store
+                await _indexWorkitemStore
+                    .UpdateWorkitemProcedureStepStateAsync(
+                        workitemMetadata,
+                        watermarkEntry.Value.NextWatermark,
+                        targetProcedureStepStateStringValue,
+                        cancellationToken)
+                    .ConfigureAwait(false);
+
+                // Delete the blob with the old watermark
+                await TryDeleteWorkitemBlobAsync(identifier, watermarkEntry.Value.CurrentWatermark, cancellationToken)
```

Comment:
So if an exception occurs here, then we will have duplicate files. What is the plan for deleting those files. How can we know the list of files to be deleted?

Reply:
Eventually we will need a scheduled job to run in the background that'll cleanup the orphan blobs (based on the WorkitemKey and Watermark in the file name)